Enemy XP, Upgrade UI, and Algorithms

NamesAreHard6600 opened the pull request:
Adds an Upgrade UI and allows you to select an upgrade from 3 choices, picked using the bucket system. 
Testing: 
Ensure the portal is set to Lust
Enter the game, and kill a few enemies to get XP
Ensure the game pauses and the upgrade UI shows up. 
Select and upgrade, and ensure it applied properly.

There are very few upgrades, as I'm waiting for the character refactor before adding more. This means you can currently get the same upgrade multiple times. This either needs to be changed, or all upgrades need to be repeatable.

NamesAreHard6600 committed:
Upgrade Templating

Template upgrades for future use.

NamesAreHard6600 committed:
Merge branch 'main' into Upgrades

NamesAreHard6600 committed:
Enemies Drop XP Orbs Which Level Up Character

Enemies can drop XP which move towards the player when close, and cause a level up when certain thresholds are hit.

NamesAreHard6600 committed:
Arrows Don't Hit Orbs

NamesAreHard6600 committed:
Continue templating upgrades

Upgrades can now be placed into buckets. Selection is not complete. More accurate Upgrade class is made.

NamesAreHard6600 committed:
Merge branch 'pause-menu' into Upgrades

NamesAreHard6600 committed:
Upgrade Bucket System

Not fully tested. No choice implemented.

NamesAreHard6600 committed:
Merge branch 'main' into Upgrades

NamesAreHard6600 committed:
Upgrade UI

Implemented an Upgrade UI where you can select an upgrade from 3 choice.
Unimplemented:
UI Prettiness
Blocking Used Upgrades?
Adding Description, rarity, etc.

NamesAreHard6600 committed:
Upgrade UI Pretiness

Made the UI actually nice to look at. Add more info to the UI, like description and rarity. Not getting the same upgrade multiple times will be implemented when there's more upgrades and upgrades that don't work multiple times.

profile1code reviewed (changes requested):
The upgrade screen does not scale with the player's screen size, resulting in a very small menu when playing the game at a high resolution (like 4k). Changes needed to make this scale with screen size.

NamesAreHard6600 committed:
Fixed Window Scaling and Panels Counting as Clicks

Upgrade UI scales with resolution
Panels don't trigger the upgrade, only hitting the button does.
